haystack/core/pipeline/async_pipeline.py

```diff
@@ -212,29 +213,30 @@ async def _run_component_async(component_name: str, component_inputs: Dict[str,
                         loop = asyncio.get_running_loop()
                         outputs = await loop.run_in_executor(None, lambda: instance.run(**component_inputs))
 
-                component_visits[component_name] += 1
-
-                if not isinstance(outputs, dict):
-                    raise PipelineRuntimeError(
-                        f"Component '{component_name}' returned an invalid output type. "
-                        f"Expected a dict, but got {type(outputs).__name__} instead. "
+                    component_visits[component_name] += 1
+
+                    if not isinstance(outputs, dict):
+                        raise PipelineRuntimeError(
+                            f"Component '{component_name}' returned an invalid output type. "
+                            f"Expected a dict, but got {type(outputs).__name__} instead. "
+                        )
+
+                    span.set_tag("haystack.component.visits", component_visits[component_name])
+                    span.set_content_tag("haystack.component.outputs", deepcopy(outputs))
```

Comment:
We were setting the tags when the context manager already exited. Not related to the bug.

Reply:
Could we separate this into a separate PR so we could get in faster without waiting for a review on everything else?

Reply:
https://github.com/deepset-ai/haystack/pull/8911